requests-for-comments/2024-07-01-alerting.md

```diff
@@ -0,0 +1,160 @@
+- `anomaly_condition`: Copy of the condition for the alert in case original got updated.
+- `threshold_met`: Whether the condition was met.
+- `notification_sent`: Whether notification was sent.
+- `error_message`: Any errors encountered.
+
+This model provides detailed logging for debugging and auditing purposes. It gives accountability and transparency to 
+users by recording the results of each alert evaluation.
+Users may ask why an alert was triggered and if the system is working as expected. The `AlertCheck` model helps answer 
+these questions by providing a detailed history of alert evaluations.
+
+We might consider discarding old `AlertCheck` records after a certain period to manage database size.
+
+#### Task Execution with Celery
+
+##### Why Use Groups, Chains, and Chunks
+
+**Groups**: Allow multiple tasks to be executed in parallel, providing the ability to run concurrent checks for 
+different alerts. By creating groups of tasks, we can limit the number of parallel operations to control system load 
+and ensure efficient resource utilization.
+
+**Chains**: Enable sequential execution of tasks, ensuring that each alert check is followed by its corresponding 
+notification task. This sequential execution is crucial for maintaining the logical flow of operations, ensuring that 
+notifications are only sent after checks are completed.
+
+**Chunks**: Divide large lists of tasks into smaller, more manageable chunks. This method is particularly useful for 
+handling bulk operations without overwhelming the system. Chunks ensure that large sets of alert checks are processed 
+in smaller batches, which are then sequentially handled by chains within groups. Celery will make sure to not launch a 
+full task for each item in the list, but rather a task for each chunk.
+
+Combining these three constructs (groups, chains, and chunks) provides a robust framework for managing the complexity 
+of alert checks and notifications. This approach ensures scalability, efficiency, and maintainability, making it 
+suitable for handling potentially large volumes of alerts and checks in PostHog.
+
+##### Check Alert Task
+
+The `check_alert` task evaluates an alert condition and logs the result. If the condition is met, it triggers a 
+notification task.
+
+Rough code outline:
+
+```python
+from celery import shared_task, chain
+from .models import Alert, AlertCheck
+from django.utils import timezone
+
+@shared_task # Think about expiration time and timeout
+def check_alert(alert_id):
+    alert = Alert.objects.get(id=alert_id)
+    
+    # Keep in mind idempotency in case of retries - e.g., check if check is already in the database for this interval
+
+    calculated_value = 42  # Example calculated value - hand off calculation to existing insight code
+    threshold_met = True  # Example threshold check
+
+    alert_check = AlertCheck.objects.create(
+        alert=alert,
+        calculated_value=calculated_value,
+        threshold_met=threshold_met,
```

Comment:
since we're adding thresholds... i'd love to be able to add thresholds in the UI that don't alert just so I can get some nice visuals on insights without needing alerts

that has the benefit we can ship it as soon as its made and start benefitting and getting user feedback on the UX

Reply:
yep, i want to see the threshold on the graph because that's useful but don't need an alert for every threshold

e.g. measuring web vitals, i want to colour the graphs because its hard to remember the 4 different thresholds (well, 8 because they have good, medium, poor) but I don't need an alert when they change cos they don't always indicate something i need to immediately react to